### PR TITLE
Hash entry rework 2 [Work in progress]

### PR DESCRIPTION
**Pull request by skiminki**
This series of patches adds 8 more bits per hash entry, reducing the number of bad TT fetches significantly.

There is an nps drop of couple % points. Some of that could be gained back by optimizations.

### CODE REVIEW

**Comment by skiminki** [comment]
On my box, I'm seeing a 2-3% nps drop, which is probably the reason why fishtest is going to fail with ~10 Elo.

The reason why I'm tentatively hopeful for this work is that I'm seeing some amount of change in the `stockfish bench` output. The most prominent is position 44:

**Stockfish vanilla:**
```
Position: 44/47
info depth 1 seldepth 1 multipv 1 score cp 1955 nodes 66 nps 66000 tbhits 0 time 1 pv h4h5 d5a2
info depth 2 seldepth 2 multipv 1 score cp 1604 nodes 254 nps 254000 tbhits 0 time 1 pv h4h5 d5f5
info depth 3 seldepth 4 multipv 1 score cp 953 nodes 390 nps 390000 tbhits 0 time 1 pv e4c5 e7c5 d4c5 d5c5
info depth 4 seldepth 4 multipv 1 score cp 928 nodes 533 nps 533000 tbhits 0 time 1 pv c3c4 f7e8 c4d5 d8d5
info depth 5 seldepth 5 multipv 1 score cp 883 nodes 669 nps 669000 tbhits 0 time 1 pv c3c4 f7e8 c4d5 d8d5 h4h5
info depth 6 seldepth 6 multipv 1 score cp 770 nodes 1144 nps 572000 tbhits 0 time 2 pv c3c4 f7e8 c4d5 e6d5 g3c3 d5e4
info depth 7 seldepth 10 multipv 1 score cp 928 nodes 1456 nps 728000 tbhits 0 time 2 pv c3c4 f7e8 c4d5 d8d5
info depth 8 seldepth 10 multipv 1 score cp 935 nodes 2038 nps 1019000 tbhits 0 time 2 pv c3c4 f7e8 c4d5 d8d5 a2a3 d5d4 g4h5 e8d7 h5h6
info depth 9 seldepth 13 multipv 1 score cp 962 nodes 7477 nps 1869250 tbhits 0 time 4 pv c3c4 f7e8 c4d5 d8d5 g3d3 a8b8 g2g3
info depth 10 seldepth 16 multipv 1 score cp 642 nodes 21013 nps 2334777 tbhits 0 time 9 pv c3c4 d5e4 g4e4 e7f6 e4f3 f7e7 g3h3 f8d7 g2g4 c6c5 f3b7 f6d4
info depth 11 seldepth 17 multipv 1 score cp 821 nodes 23011 nps 2301100 tbhits 0 time 10 pv c3c4 d5e4 g4e4 e7f6 g3f3
info depth 12 seldepth 18 multipv 1 score cp 851 nodes 27716 nps 2132000 tbhits 0 time 13 pv c3c4 d5e4 g4e4 e7f6 f1b1 f7e7 b1b7 d8d7 e4c6 a8e8 b7d7 f8d7 d4d5 e7d8 d5e6
info depth 13 seldepth 22 multipv 1 score mate 4 nodes 41377 nps 2177736 tbhits 0 time 19 pv g4g7 f7e8 g7e7 e8e7 g3g7 e7e8 e4f6
bestmove g4g7 ponder f7e8
```

With these patches:
```
Position: 44/47
info depth 1 seldepth 1 multipv 1 score cp 1955 nodes 59 nps 59000 tbhits 0 time 1 pv h4h5 d5a2
info depth 2 seldepth 2 multipv 1 score cp 1955 nodes 154 nps 154000 tbhits 0 time 1 pv h4h5 d5a2
info depth 3 seldepth 3 multipv 1 score cp 2130 nodes 221 nps 221000 tbhits 0 time 1 pv a2a3 e7a3 h4h5
info depth 4 seldepth 4 multipv 1 score cp 2120 nodes 659 nps 329500 tbhits 0 time 2 pv g1h2 d5a2 h4h5
info depth 5 seldepth 5 multipv 1 score cp 2096 nodes 1118 nps 559000 tbhits 0 time 2 pv h4h5 d5a2
info depth 6 seldepth 10 multipv 1 score cp 1438 nodes 2680 nps 1340000 tbhits 0 time 2 pv f1e1 e7f6 e4f6 f7e7 g4g7 e7d6 g7b7
info depth 7 seldepth 10 multipv 1 score cp 1535 nodes 2869 nps 1434500 tbhits 0 time 2 pv f1e1 e7f6 e4f6 f7e7 g4g7 e7d6
info depth 8 seldepth 12 multipv 1 score cp 1156 nodes 5442 nps 1814000 tbhits 0 time 3 pv h4h5 e7f6 g4g8 f7e8 e4f6 e8e7 f6d5 c6d5 g3g7 e7d6 g7b7 e6e5 d4e5 d6e5
info depth 9 seldepth 17 multipv 1 score mate 4 nodes 6697 nps 1674250 tbhits 0 time 4 pv g4g7 f7e8 g7e7 e8e7 g3g7 e7e8 e4f6
info depth 10 seldepth 8 multipv 1 score mate 4 nodes 6878 nps 1719500 tbhits 0 time 4 pv g4g7 f7e8 g7e7 e8e7 g3g7 e7e8 e4f6
info depth 11 seldepth 8 multipv 1 score mate 4 nodes 7060 nps 1765000 tbhits 0 time 4 pv g4g7 f7e8 g7e7 e8e7 g3g7 e7e8 e4f6
info depth 12 seldepth 8 multipv 1 score mate 4 nodes 7313 nps 1828250 tbhits 0 time 4 pv g4g7 f7e8 g7e7 e8e7 g3g7 e7e8 e4f6
info depth 13 seldepth 8 multipv 1 score mate 4 nodes 7672 nps 1918000 tbhits 0 time 4 pv g4g7 f7e8 g7e7 e8e7 g3g7 e7e8 e4f6
bestmove g4g7 ponder f7e8
```

With these patches, the mate-in-4 is found significantly faster, and I would guess that the reason is a bad TT fetch or two in vanilla SF, misdirecting the search.

I do realize that the bench is not perhaps the best measure of quality here, but at least it's promising. I think it would make sense to try out next some difficult positions where SF struggles, to see whether these patches improve anything at all. Anyways, for analysis (at least for me), the 2-3% slowdown wouldn't be of any significance at all if the search quality improves even a bit. But, it still needs to be shown that the search quality actually improves. I don't know about that yet.

The original motivation for these patches was the realization that 3/65536 (= 0.0046%) fetches return a garbage TT hit when the TT is full. Garbage means an undetected hit from an incorrect position, and this is regardless of the hash size. It may not feel like a big number, but on modern machines, those collisions happen in high frequencies.

I'm also aware of a decades old study which claims that hash collisions don't affect things significantly. However, the computers today are maybe 100x faster than what they were when those studies were made. So, I think a refresh is warranted.

**Comment by skiminki** [comment]
I ran the test position in https://github.com/official-stockfish/Stockfish/issues/2169 with SF vanilla and my patch. I used single-threaded search with 3GB hash (I should have used 2GB to match the experiment in the issue...). But anyways, the results are interesting:

cpgraph

In this run, the better hash collision detection results in significantly faster CP eval increase vs the nodes.

The UCI commands:

```
setoption name Hash value 3072
ucinewgame
setoption name UCI_AnalyseMode value true
position fen 8/7b/3kp3/4p3/1P4P1/2K5/4B3/8 w - - 0 1
isready
go depth 70
```

The UCI logs:
[sf-hash24.txt](https://github.com/official-stockfish/Stockfish/files/4809312/sf-hash24.txt)
[sf-vanilla.txt](https://github.com/official-stockfish/Stockfish/files/4809313/sf-vanilla.txt) (up to depth 69 only)

**Comment by vondele** [comment]
@skiminki the ultimate test will be fishtest, as always things are a balance between speed and accuracy. Revisiting old results is very good however, things have changed a lot. 

My experience so far is that one has to be careful judging results on single positions, any change to search can have big effects on single positions. 

Similarly, the number of nodes reported by bench can vary wildly, even for the smallest changes, 20% change is common. It is not a really good indicator of improvement of single patches... there is a trend of course in the long run.

**Comment by skiminki** [comment]
Thanks, I'm aware of that and I plan to run more testing. But I don't have a collection of good test positions, so I'm hoping that others would help me here.

I would expect that there is a break-even when the reduced number of bad hash fetches outweighs the nps drop. But I don't know yet that at what TC and hash size configs that break-even would be.

File: src/tt.cpp [hunk 1]
```diff
@@ -28,30 +28,202 @@
...
+
+          case RANK_7: // write promotion
+              promotionBits = PROMOTION | 00070U | (encodedMove & 00070) << 9;
+              ret = (encodedMove & 07707U) | promotionBits;
+              break;
+
+              // rank 4/5
+          default:
+              ret = ENPASSANT | encodedMove;
+              break;
```
Comment (Sopel97) [inline src/tt.cpp:101]: If this encoding scheme is kept this could maybe be an array `or_type[RANK_NB]` so this can be done as `encodedMove |= or_type[sq_rank]` and just special case `RANK_2` and `RANK_7`. 

For `case RANK_7:`  `& 07707U` seems redundant because it will always be ored with `00070U` and is already `&= 0xFFF`?

File: src/tt.cpp [hunk 2]
```diff
@@ -28,30 +28,202 @@
...
+      setBitField<TTEntryPacked::Gen5>(packedData.bits, TT.generation5);
+      setBitField<TTEntryPacked::Pv>(packedData.bits, pv);
+      setBitField<TTEntryPacked::Bound2>(packedData.bits, b);
+      setBitField<TTEntryPacked::Depth8>(packedData.bits, d - DEPTH_OFFSET);
+      doStore = true;
   }
+
+  // store only if we changed something to save memory BW
+  if (doStore)
+      TT.table[m_clusterIndex].entry[m_slotIndex] = packedData;
```
Comment (Sopel97) [inline src/tt.cpp:226]: `cluster.entry`?

File: src/tt.cpp [hunk 3]
```diff
@@ -113,32 +285,48 @@ void TranspositionTable::clear() {
...
+  TTEntryPacked* const tte = &cluster.entry[0];
+  const uint32_t key = entryKeyFromZobrist(fullKey);
+  const uint8_t extraKey = extraKeyFromZobrist(fullKey);
 
   for (int i = 0; i < ClusterSize; ++i)
-      if (!tte[i].key16 || tte[i].key16 == key16)
+  {
+      const uint32_t tteKey = getClusterEntryKey(cluster.keys, i);
+      const uint32_t tteExtraKey = extractBitField<TTEntryPacked::ExtraHash3>(tte[i].bits);
+      const bool occupied = (extractBitField<TTEntryPacked::Move13>(tte[i].bits) != 0);
```
Comment (Sopel97) [inline src/tt.cpp:300]: this could use a function to extract raw bits of a given bitfield, i.e. without shifting, just apply mask.

File: src/tt.cpp [hunk 4]
```diff
@@ -113,32 +285,48 @@ void TranspositionTable::clear() {
...
+
+      if (!occupied || (tteKey == key && tteExtraKey == extraKey))
```
Comment (Sopel97) [inline src/tt.cpp:302]: could be more lazy with hash comparison? Could help with small searches.

File: src/tt.h [hunk 5]
```diff
@@ -24,36 +24,57 @@
...
+struct TTEntryPacked {
+
+  using Move13     = BitFieldDesc< 0U, 13U, uint16_t>;
+  using ExtraHash3 = BitFieldDesc<13U,  3U, uint8_t>;
+  using Value16    = BitFieldDesc<16U, 16U, Value, true>; // sign-extend
+  using Eval16     = BitFieldDesc<32U, 16U, Value, true>; // sign-extend
+  using Gen5       = BitFieldDesc<48U,  5U, uint8_t>;
+  using Pv         = BitFieldDesc<53U,  1U, bool>;
+  using Bound2     = BitFieldDesc<54U,  2U, Bound>;
+  using Depth8     = BitFieldDesc<56U,  8U, uint8_t>;
```
Comment (Sopel97) [inline src/tt.h:47]: this could use some static_assert to check whether the fields are disjoint.

Comment (skiminki) [reply src/tt.h:47]: Thanks for the reviews! I'll address these one way or another on my next PR iteration

**Comment by vondele** [comment]
I have attached some scripts and data to test performance on mate finding. I see no difference so far, but maybe the script is useful for development :
```
$ python3 matecheck.py --stockfish ./stockfish.master --nodes 10000000 && python3 matecheck.py --stockfish ./stockfish.hash --nodes 10000000
Using ./stockfish.master with 10000000 nodes
Total fens:    6566
Found mates:   4997
Best mates:    3974
Better mates:  0
Using ./stockfish.hash with 10000000 nodes
Total fens:    6566
Found mates:   4983
Best mates:    3971
Better mates:  0
```
[matecheck.tar.gz](https://github.com/official-stockfish/Stockfish/files/4809953/matecheck.tar.gz)

**Comment by skiminki** [comment]
This is a mate-in-19 position from https://tcec-chess.com/#div=sf&game=6&season=18

```
setoption name Hash value 8192
ucinewgame
setoption name UCI_AnalyseMode value true
position fen 8/4k2P/4P3/5P2/5K2/R3n2r/8/8 w - - 5 65
isready
go infinite
```

[
cpgraph
](url)

Mate translated to score 250.

**Comment by skiminki** [comment]
TCEC S18 SuFi G17 book exit

```
setoption name Hash value 16384
ucinewgame
setoption name UCI_AnalyseMode value true
position fen r1b1kb1r/2q2p1p/p2ppP2/1pn5/3NP3/2N2Q2/PPP4P/2KR1BR1 b kq - 2 16
isready
go infinite
```

cpgraph

Hash24 has faster convergence to 0.00

**Comment by skiminki** [comment]
From https://tcec-chess.com/#div=sf&game=34&season=18 after 75... Rg6
SF originally thinks white is winning or considerably better, but the position is a draw due to no progress before 50 mr.

```
setoption name Hash value 8192
setoption name SyzygyPath value /mnt/data-n0/syzygy
setoption name UCI_AnalyseMode value true
ucinewgame
position fen 8/6nk/6r1/4P2p/4Q2P/1K6/8/8 w - - 53 76
isready
go infinite
```

cpgraph

Here, vanilla actually hits 0.00 earlier whereas hash24 keeps low non-zero scores for a while. I would guess that vanilla wins due to a fortunate hash collision cutting some search branches. But regardless, a win for vanilla.

**Comment by vondele** [comment]
@skiminki testing with specific positions is flawed... even the tiniest change to search or eval can lead to completely different results for the search on a couple of positions. It is one of the major strengths of fishtest, to put everything on a sound statistical basis.

**Comment by Alayan-stk-2** [comment]
Maybe run a fixed games LTC with nodestime to test hash24 vs hash16 at equivalent nps.

If at equal nps 60s+0.6 40K games you can get a decent likelihood of superiority, then :
- Work on speed optimization will be justified
- You could argue for a VLTC test, as presumably bigger hash and longer TC would make this more significant and add more data to support a merge even at the cost of some slowdown.

If a nodestime test fails to get a good LoS, then either this is not useful enough for SF in the general case, or fishtest is too far removed from analysis-TC/hw issues.

**Comment by skiminki** [comment]
I'm going to run couple more test positions to see whether more accurate hashing is better in long searches. After that I think it would be a time for a discussion on whether the upstream is interested on these patches or not, and how to proceed. Consider these postings on searches as circumstantial evidence, not something that I'd use to get these patches merged. See below for rationale.

> testing with specific positions is flawed... 

Unfortunately, I don't have the computing power to start running testing on something like rapid TC or TCEC SuFi TC, so individual test positions is the best that I can do.

So far it seems that the search with these patches converges better after millions or billions of nodes. This means that we would probably need to run much much longer TC than what is usually done in fishtest for quality improvements. In terms of pure Elo, I think it is entirely possible that more accurate hashing is never worth the nps loss in fishtest, but the same might not be true in TCEC time controls.

Also, I'm personally using SF for analysis and analysis only. If the evaluations converge faster with my patches for me, then I'm happy. I'd be even happier if SF sometimes finds better moves that otherwise would be hidden behind hash collisions. I don't know whether this can actually be the case, or whether the search would eventually always work around the hash collisions given enough time.

I personally don't care at all whether the patches would drop Elo in short TC. This said, I do acknowledge that others have different goals with SF. This means I'm not necessarily expecting to get these patches merged in the upstream, in which case this work will remain in my own fork. (Similar things happened also with the init speed improvements etc. I do understand that not everyone is as interested in the SF launch speed at the cost of some code complexity, even if the launch speed is quite important for me.)

> Maybe run a fixed games LTC with nodestime to test hash24 vs hash16 at equivalent nps.
> If at equal nps 60s+0.6 40K games you can get a decent likelihood of superiority, then...

That would make sense. Although, I'm not quite sure yet whether this TC would be enough. There's basically two clear improvements with my patches, but they'll only start to be effective with higher node counts:

- better hash collision detection (generally we need hash to be somewhat occupied before this kicks in). Mostly empty hash = very little collisions.
- the ability to fetch hash entries with entry key 0. (Current upstream probe() will always report ttHit=false for such entries). Again, this needs some number of nodes in searches to really start kicking in.

Anyway, I need to think about this.

> If a nodestime test fails to get a good LoS, then either this is not useful enough for SF in the general case, or fishtest is too far removed from analysis-TC/hw issues.

It is also possible that:
- there is a bug in my patches. This stuff in SF is actually relatively complex, to me at least (I'm still working with verification)
- SF search with low nodes is inadvertently tuned for some amount of hash collisions, and particularly, for the inability to use hash16==0 TT entries.

Anyway, I need to think about this, too. (I think it's possible that there's some reliance on hash collisions, given the difficulties to get mstembera's "Use 128 bit multiply for TT index" patch passing in fishtest. This patch changes hash collision patterns.)

> Work on speed optimization will be justified

I'm actually already working on that in my own fork and I've been able to recover about 0.5 %-points of the nps loss. I haven't been pushing the updated patches here yet to avoid unnecessary spam. But if interested, the current patches are found in https://github.com/skiminki/Stockfish (master branch).

**Comment by gonzalezjo** [comment]
You could try an extremely-low priority, very long time control patch with nodestime. For example, 180+1.8s. It will take forever to get any meaningful data, but (1) it won't bog up fishtest, and (2) late data is better than no data.

**Comment by NKONSTANTAKIS** [comment]
I am curious if this helps with that CCC bullet slowdown. We should send it to greco for a fast test.

**Comment by skiminki** [comment]
@NKONSTANTAKIS I don't think so. I thought the CCC bullet slowdown was already resolved? (Possibly something related to NUMA memory migration, IIRC)

**Comment by skiminki** [comment]
I have two more search studies, and I think these will be the final ones:

**Study 1** https://tcec-chess.com/#div=sf&game=65&season=18 Will SF allow the Qxb4 queen sac?

cpgraph

SF-vanilla will play 20... Nxb4 until 227.6G nodes. At that point, SF-vanilla switches to 20... a6.

SF-hash24 will play 20... Nxb4 until 67.9G nodes. At that point, SF-hash24 switches to 20... a6. At 70.0G nodes, SF-hash24 switches to 20... a5 (Leela's expectation).  There is some switching between 20... a6/a5/Nxb4 until 238G nodes.

Logs:
- [sf-vanilla.txt](https://github.com/official-stockfish/Stockfish/files/4858941/sf-vanilla.txt)
- [sf-hash24.txt](https://github.com/official-stockfish/Stockfish/files/4858940/sf-hash24.txt)

**Study 2** https://tcec-chess.com/#div=sf&game=60&season=18 What will SF play for move 4w?

cpgraph

SF-vanilla alternates between 4. c3 and 4. Be3 after 1M nodes.

SF-hash24 would play 4. c3 after 1M nodes, and switches to 4. d5 (Leela's move) after 2.9G nodes. There's a brief lapse back to 4. c3 from 5.4Gn to 6.3Gn, then switching back to 4. d5 until the search end (35Gn)

Log:
- [sf-vanilla.txt](https://github.com/official-stockfish/Stockfish/files/4859015/sf-vanilla.txt)
- [sf-hash24.txt](https://github.com/official-stockfish/Stockfish/files/4859014/sf-hash24.txt)


Based on the positions I've searched (there's couple more but won't give any new info), the CP evaluation converges faster with hash24 patches than vanilla, and hash24 may also play a bit stronger moves than vanilla. I think there are now somewhat strong indications that the current hashing scheme gets in the way of search with high enough node counts, and there might be some actual strength to be gained. Ofc, this needs to be shown in proper play testing.

> You could try an extremely-low priority, very long time control patch with nodestime. For example, 180+1.8s. 

Yeah, maybe something like this would now make sense. My plan for these patches is now as follows:
- rebase the patches and update this PR. I may still look into adding some extra protection for concurrent races based on the 3 extra bits in the hash entry. Hopefully this will be close to zero overhead. (I want to prevent a partial move-only hash update for a wrong position in store())
- either run some local testing or submit nodestime test in fishtest

I should be able to get back to this in couple of days.

**Comment by skiminki** [comment]
Local match between:

- sf-vanilla: https://github.com/skiminki/Stockfish/commit/87ebe9d7e620b0a7d1a9ac5d8987ecc9015ca89b (this is essentially upstream at https://github.com/skiminki/Stockfish/commit/67818ee9481ba99369fa8a8d92e5c50428fb300e)
- sf-hash24: https://github.com/skiminki/Stockfish/commit/c20453bc77c3a4d058094b09e714ea97035bbd6e
 
TC 180s+1.8s, hash 128 MB, threads=1, games 500, noob_3moves.epd, 6-men syzygy
HW core-i7 8700k / 64 GB memory / fast mSATA SSD

Cutechess launch:
```
cutechess-cli \
        -draw movenumber=40 movecount=10 score=10 \
        -tb /mnt/data-n0/syzygy -tbpieces 6 \
        -openings file=noob_3moves.epd format=epd order=random \
        -pgnout match.pgn \
        -concurrency 5 \
        -rounds 500 \
        -engine conf=stockfish-vanilla -engine conf=stockfish-hash24 \
        -each tc=180+1.8 dir=/tmp/cutechess
```

Engine options for both:

```
        "restart": "on",
        "options": [
            {
                "name": "Hash",
                "value": "128"
            },
            {
                "name": "SyzygyPath",
                "value": "/mnt/data-n0/syzygy"
            }
        ],
        "protocol": "uci",
        "workingDirectory": "/tmp/cutechess/sf0"
```

```
Score of stockfish-vanilla vs stockfish-hash24: 30 - 36 - 434  [0.494] 500
...      stockfish-vanilla playing White: 19 - 14 - 217  [0.510] 250
...      stockfish-vanilla playing Black: 11 - 22 - 217  [0.478] 250
...      White vs Black: 41 - 25 - 434  [0.516] 500
Elo difference: -4.2 +/- 11.0, LOS: 23.0 %, DrawRatio: 86.8 %
```

Ok, not exactly decisive, but 180s+1.8s seems to be on the better side of the break-even where the more precise search is more important than the nps loss, on my box at least. I ran earlier a preliminary 60s+0.6s / 64 MB match. That was dead even at 200 games.

I'll run another test with 360s+3.6s / 256 MB hash / 500 games. That'll probably take a day or so. Provided it yields similar results, I'll do a final optimization pass for the patches and prepare a fishtest run.

**Comment by skiminki** [comment]
Result of TC 360s+3.6s, hash 256 MB, threads=1, games 500, noob_3moves.epd, 6-men syzygy

```
Score of stockfish-hash24 vs stockfish-vanilla: 40 - 39 - 421  [0.501] 500
...      stockfish-hash24 playing White: 24 - 20 - 206  [0.508] 250
...      stockfish-hash24 playing Black: 16 - 19 - 215  [0.494] 250
...      White vs Black: 43 - 36 - 421  [0.507] 500
Elo difference: 0.7 +/- 12.1, LOS: 54.5 %, DrawRatio: 84.2 %
Finished match```

That's negligible victory for hash24. I think I'm going to try with a smaller hash/TC ratio next.

**Comment by skiminki** [comment]
Result of TC 180s+1.8s, hash 64 MB, threads=1, games 500, noob_3moves.epd, 6-men syzygy 

```
cutechess-cli \
        -tb /mnt/data-n0/syzygy -tbpieces 6 \
        -openings file=noob_3moves.epd format=epd order=random \
        -pgnout match.pgn \
        -concurrency 6 \
        -rounds 500 \
        -srand 3 \
        -engine conf=stockfish-hash24 -engine conf=stockfish-vanilla \
        -each tc=180+1.8 dir=/tmp/cutechess
```

```
Score of stockfish-hash24 vs stockfish-vanilla: 44 - 27 - 429  [0.517] 500
...      stockfish-hash24 playing White: 25 - 11 - 214  [0.528] 250
...      stockfish-hash24 playing Black: 19 - 16 - 215  [0.506] 250
...      White vs Black: 41 - 30 - 429  [0.511] 500
Elo difference: 11.8 +/- 11.4, LOS: 97.8 %, DrawRatio: 85.8 %
Finished match
```

With the smaller hash, this time the results are more convincing for hash24, and the LOS value would indicate that the results are no longer in small sample size territory.

I'm currently thinking that the hash24 strength gain/loss is related to how much the hash is overwritten per move, because that essentially would relate to the harmful collision frequency. (Collisions with earlier searches shouldn't be that harmful because there's the aging factor which reduces the effective depth of the TT entry.)

In other words, the bigger the hash compared to nodes/move, the less it matters that we're not detecting collisions that precisely.

Anyways, this may be a problem for getting a pass in standard 1-threaded fishtests, because the hash/TC ratio there is quite big, certainly bigger than what I have been running here with positive results.

Continuing, in TCEC or in long analysis, I would think the full hash rewrites per move is way above 1. SF search in TCEC regularly goes into dozens of giganodes and the hash size is 64 GB. 64 GB translates roughly to 6.4 billion stored positions. SF runs roughly at 150..200 Mnps in TCEC (depending on the position), so nodes accumulate quickly.

Maybe I'll run a multi-threaded test next.

**Comment by vondele** [comment]
@skiminki the hash size can be used as a parameter on fishtest, so one can easily run e.g. an STC test with 1MB hash, this is much easier than running at VLTC. This has been done in the past for features that were considered sensitive to hash pressure.

**Comment by skiminki** [comment]
Hmmm... maybe that could be the next option. Anyways, I think I'll still run some kind of SMP test locally to ensure that we're not introducing some pathological race conditions.

**Comment by vondele** [comment]
I realized there is a quick check we can do to measure the impact of the key length. With little overhead, we can remove bits from the key and measure the impact. Two tests running remove 4 and 2 bits respectively:
https://tests.stockfishchess.org/tests/view/5f062b4f59f6f03532894665
https://tests.stockfishchess.org/tests/view/5f0633e359f6f03532894672

**Comment by skiminki** [comment]
@vondele  I think there's not enough hash pressure with those 2 tests to show significant Elo drops due to increased collisions. We should see more significant Elo drops with smaller hashes.

**Comment by skiminki** [comment]
I just realized that I messed up with the previous test. Oh well... But anyways, what I'm trying to figure out with local testing is that how much hash pressure there has to be before there's significant strength impact. I'll start running some STC SMP testing.

**Comment by vondele** [comment]
@skiminki we can rerun one of the tests with smaller hash. However, note that the hash will (after the first few moves in a game) always be full with positions, just possibly of an older generation. So, the hash pressure idea right now is still more of a hypothesis (which can be tested). Anyway, right now it seems that even using just 12 bits for the key doesn't lead to a significant Elo drop, which is quite interesting.

**Comment by mstembera** [comment]
FYI This is rather old but it was verified in the past that even 12 bit keys are fine.  At 10 bits we start to see a drop off.
https://tests.stockfishchess.org/tests/view/5431cc620ebc597e655586ca
https://tests.stockfishchess.org/tests/view/54355f180ebc593533cac2aa

**Comment by vondele** [comment]
OK, nice, they are much in line with what we find now (6 years later). They do use 2Mb hash, I see and are at 15+0.05s.

**Comment by skiminki** [comment]
> FYI This is rather old but it was verified in the past that even 12 bit keys are fine. At 10 bits we start to see a drop off.
> OK, nice, they are much in line with what we find now (6 years later). They do use 2Mb hash, I see and are at 15+0.05s.

Right. Anyways, I'm not quite convinced on how these STC results scale to TCEC-style long time control or analysis. That's why I started by analyzing some positions from the previous TCEC SuFi with tens of billions of nodes, and hash24 performs there consistently better, as is seen in those graphs I posted earlier.

There's also another anomaly that hash24 fixes: The current SF vanilla is bad at figuring out the M4/M5 positions in the default bench but hash24 finds both of them and finds them fast. This has been consistent over rebases. This may not be an actual problem, since the hash is supposed to be cleared when you start analyzing a different game. Bench doesn't do those clears. But since the bench is just a bench, maybe this is not a problem that even needs fixing.

I now have the first SMP results:

Result of TC 5+0.05, hash 8 MB, threads=12, games 500, noob_3moves.epd, 6-men syzygy

```
Score of stockfish-hash24 vs stockfish-vanilla: 85 - 76 - 339  [0.509] 500
...      stockfish-hash24 playing White: 44 - 39 - 167  [0.510] 250
...      stockfish-hash24 playing Black: 41 - 37 - 172  [0.508] 250
...      White vs Black: 81 - 80 - 339  [0.501] 500
Elo difference: 6.3 +/- 17.3, LOS: 76.1 %, DrawRatio: 67.8 %
```

> So, the hash pressure idea right now is still more of a hypothesis (which can be tested). Anyway, right now it seems that even using just 12 bits for the key doesn't lead to a significant Elo drop, which is quite interesting.

I personally am somewhat convinced that we need more hash bits with longer time controls, but TBH, it doesn't mean I'm right. I also don't have a solid hypothesis why this is the case, and I still don't have good data on when more hash bits becomes a win over the nps drop. I only have circumstantial data that more hash bits _should_ be a win.

Other than hash pressure, the SMP result above can also be about:
- small sample size
- less concurrent collisions/races (the way hash24 accesses the hash should be less prone to concurrency races than vanilla)

But again, it's circumstantial data that points in the direction to more hash bits and/or more atomic hash access is a win. Especially considering that hash24 is that ~2% slower than vanilla. The slowdown alone should amount to a loss of couple Elo points.

**Comment by vondele** [comment]
@skiminki I'll close this WIP PR, please reopen it after rebasing if you plan to continue it.

**Comment by skiminki** [comment]
Just a quick update. I started running tests on another box (6C/12T @ 3.2 GHz, a recent Intel NUC box). There I'm having less favorable STC results, probably because the nps hit is steeper. I'd speculate that the steeper nps hit is due to better memory latency to CPU speed ratio on that box, disfavoring the extra computation more.

I'll probably pick up this task again for NNUE in the coming weeks. Perhaps the added per-TT-access computation matters less with the much heavier evaluation of NNUE, but we'll see.

(In case anyone's interested, I have rebased these patches for NNUE on my own branch.)

**Comment by skiminki** [comment]
I decided to extract the simple fix for key16==0 being always a miss from the hash24 patch set: https://github.com/skiminki/Stockfish/tree/zero-hash16-fix . Local testing suggests that it might be Elo neutral or slightly Elo positive for 10s+0.1s with small hash sizes.

Essentially, that "ttHit = (key==0)" condition in TT probe means that 1/65536 positions won't ever get fetched from the TT hash. Not sure how important that is for different time controls.

Fish test with regression bounds for the above is pending: https://tests.stockfishchess.org/tests/view/5f3d7e66a95672ddd56c640e . Please let me know if I should have used the standard bounds instead. If the fish test passes, I'll send a separate PR for it.

While doing this fix against upstream and comparing the bench numbers to my hash24 patches with hash truncated to 16 bits, I found a small discrepancy in bench ids and, consequently, a small bug in my hash24 patches. Perhaps there is still hope for hash24. I'll continue nodestime testing with this new fix. In any case, I plan to submit test results once they're done for future reference.

**Comment by vondele** [comment]
that would need Elo gainer bounds, since it adds complexity to deal with a special case, which I would expect not to matter much.

**Comment by skiminki** [comment]
Ok, someone already approved the test. Let me cancel it and submit new one.

**Comment by skiminki** [comment]
The STC 10s+0.1s test came back as weak reject: https://tests.stockfishchess.org/html/live_elo.html?5f3d8c74a95672ddd56c641a

```
  LLR | -2.98 [-2.94,2.94] (rejected)
   -- | --
  Elo | -0.67 [-2.72,1.46] (95%)
  LOS | 26.5%
Games | 23648 [w:10.4%, l:10.7%, d:78.9%]
```

Would it be justified to submit a LTC test to see whether there's scaling effects?

**Comment by vondele** [comment]
rather than LTC, I think you want to change once default hash. That's presumably more interesting.

**Comment by skiminki** [comment]
> rather than LTC, I think you want to change once default hash. That's presumably more interesting.

Suggestions for the hash size? Double the default hash size? Half the default hash size? The inability to store 1/65536 positions into hash (well, technically, the inability to fetch) is unrelated to the hash size.

(It is possible that plugging this hole could be visible only in endgames with lots of transpositions where some key positions happen to have hash16==0. For all I know, it could be this part of my hash24 patches that made the endgames converge faster in my preliminary hours-long analysis tests, not actually increasing the entry key to 24 bits. I don't know, though. The search interaction with hash seems quite unintuitive to me.)

**Comment by vondele** [comment]
OK, so if it is independent of hash size it is probably not worth it. I would have guessed that this would have more effect for smaller hash. If you really want you can do a speculative LTC at reduced throughput. Generally this doesn't help. Also, with NNUE, effectively we're currently playing at a large TC (scaling with nps), and we are at a level of play >> LTC from a few weeks ago.

**Comment by skiminki** [comment]
> If you really want you can do a speculative LTC at reduced throughput. Generally this doesn't help. 

I agree that this is speculative, but I think it is still worth finding out. Although I fear that even this might require way longer searches than fishtest generally does. If we're still close to +/-0 Elo with LTC, maybe I'll reconsider running the study positions again to find out whether it was 24-bit entry hash or the hash16==0 fix that improved scaling on those positions. 

> Also, with NNUE, effectively we're currently playing at a large TC (scaling with nps), and we are at a level of play >> LTC from a few weeks ago.

I would think that the hash efficiency/correctness is more connected to the actual number of searched nodes and NNUE/handcrafter eval is a bit orthogonal if not accounting for the nps changes.
